### DISCUSSION TITLE
Recommendation for Unit Testing?

### DISCUSSION THREAD

**Discussion by daners88** [ArthurSonzogni/FTXUI#644] [Q&A] [answered]
Hi, I'm hoping to use FTXUI for a project of mine. I'm looking in to how I can unit test the interactive GUI and capture mouse events, click events, etc... Is there a recommendation for what to use to accomplish this? End goal is fairly thorough code coverage. Thank you!

**Comment by ArthurSonzogni** [comment, marked as answer]
FTXUI should be quite easy to test, because you can send arbitrary events to a component, and check how it ends up affecting the UI state. The UI state is mainly stored from outside of FTXUI.

You can take example from pre-existing FTXUI tests:
https://github.com/ArthurSonzogni/FTXUI/blob/main/src/ftxui/component/input_test.cpp

It boiles down to:
```cpp
  State myState;
  Component myComponent = Mycomponent(&myState);

  myComponent->OnEvent(Event::Character("a"));
  EXPECT_EQ(myState, expected)
```

**Reply by cma0014** [reply]
What if, say, you've composed a component with multiple buttons that are wired to event handlers in a separate class (a publisher or interface call), and you want to test the button presses triggering those interface calls?
